File: app/src/main/java/org/tvheadend/tvhclient/ui/features/settings/SettingsConnectionBaseFragment.java
package org.tvheadend.tvhclient.ui.features.settings;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.afollestad.materialdialogs.MaterialDialog;

import org.tvheadend.tvhclient.MainApplication;
import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.data.repository.AppRepository;
import org.tvheadend.tvhclient.domain.entity.Connection;
import org.tvheadend.tvhclient.ui.base.callbacks.BackPressedInterface;
import org.tvheadend.tvhclient.ui.base.callbacks.ToolbarInterface;
import org.tvheadend.tvhclient.ui.base.utils.SnackbarUtils;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.preference.CheckBoxPreference;
import androidx.preference.EditTextPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

public abstract class SettingsConnectionBaseFragment extends PreferenceFragmentCompat implements BackPressedInterface, Preference.OnPreferenceChangeListener {

    @Inject
    AppRepository appRepository;

    AppCompatActivity activity;
    ToolbarInterface toolbarInterface;
    boolean connectionValuesChanged;
    Connection connection;
    ConnectionViewModel viewModel;

    private EditTextPreference namePreference;
    private EditTextPreference hostnamePreference;
    private EditTextPreference htspPortPreference;
    private EditTextPreference streamingPortPreference;
    private EditTextPreference usernamePreference;
    private EditTextPreference passwordPreference;
    private CheckBoxPreference activeEnabledPreference;
    private EditTextPreference wolMacAddressPreference;
    private EditTextPreference wolPortPreference;
    private CheckBoxPreference wolEnabledPreference;
    private CheckBoxPreference wolUseBroadcastEnabled;

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        activity = (AppCompatActivity) getActivity();
        if (activity instanceof ToolbarInterface) {
            toolbarInterface = (ToolbarInterface) activity;
        }
        MainApplication.getComponent().inject(this);

        viewModel = ViewModelProviders.of(activity).get(ConnectionViewModel.class);
        setHasOptionsMenu(true);

        // Get the connectivity preferences for later usage
        namePreference = findPreference("name");
        hostnamePreference = findPreference("hostname");
        htspPortPreference = findPreference("htsp_port");
        streamingPortPreference = findPreference("streaming_port");
        usernamePreference = findPreference("username");
        passwordPreference = findPreference("password");
        activeEnabledPreference = findPreference("active_enabled");
        wolEnabledPreference = findPreference("wol_enabled");
        wolMacAddressPreference = findPreference("wol_mac_address");
        wolPortPreference = findPreference("wol_port");
        wolUseBroadcastEnabled = findPreference("wol_broadcast_enabled");

        namePreference.setOnPreferenceChangeListener(this);
        hostnamePreference.setOnPreferenceChangeListener(this);
        htspPortPreference.setOnPreferenceChangeListener(this);
        streamingPortPreference.setOnPreferenceChangeListener(this);
        usernamePreference.setOnPreferenceChangeListener(this);
        passwordPreference.setOnPreferenceChangeListener(this);
        activeEnabledPreference.setOnPreferenceChangeListener(this);
        wolEnabledPreference.setOnPreferenceChangeListener(this);
        wolMacAddressPreference.setOnPreferenceChangeListener(this);
        wolPortPreference.setOnPreferenceChangeListener(this);
        wolUseBroadcastEnabled.setOnPreferenceChangeListener(this);
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences_add_connection, rootKey);
    }

    @Override
    public void onResume() {
        super.onResume();
        setPreferenceDefaultValues();
    }

    private void setPreferenceDefaultValues() {
        String name = connection.getName();
        namePreference.setText(name);
        namePreference.setSummary(TextUtils.isEmpty(name) ? getString(R.string.pref_name_sum) : name);

        String address = connection.getHostname();
        hostnamePreference.setText(address);
        hostnamePreference.setSummary(TextUtils.isEmpty(address) ? getString(R.string.pref_host_sum) : address);

        String port = String.valueOf(connection.getPort());
        htspPortPreference.setText(port);
        htspPortPreference.setSummary(port);

        streamingPortPreference.setText(String.valueOf(connection.getStreamingPort()));
        streamingPortPreference.setSummary(getString(R.string.pref_streaming_port_sum, connection.getStreamingPort()));

        String username = connection.getUsername();
        usernamePreference.setText(username);
        usernamePreference.setSummary(TextUtils.isEmpty(username) ? getString(R.string.pref_user_sum) : username);

        String password = connection.getPassword();
        passwordPreference.setText(password);
        passwordPreference.setSummary(TextUtils.isEmpty(password) ? getString(R.string.pref_pass_sum) : getString(R.string.pref_pass_set_sum));

        activeEnabledPreference.setChecked(connection.isActive());
        wolEnabledPreference.setChecked(connection.isWolEnabled());

        if (!connection.isWolEnabled()) {
            String macAddress = connection.getWolMacAddress();
            wolMacAddressPreference.setText(macAddress);
            wolMacAddressPreference.setSummary(TextUtils.isEmpty(macAddress) ? getString(R.string.pref_wol_address_sum) : macAddress);
            wolPortPreference.setText(String.valueOf(connection.getWolPort()));
            wolPortPreference.setSummary(getString(R.string.pref_wol_port_sum, connection.getWolPort()));
            wolUseBroadcastEnabled.setChecked(connection.isWolUseBroadcast());
        }
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.save_cancel_options_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_save:
                save();
                return true;
            case R.id.menu_cancel:
                cancel();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    protected abstract void save();

    /**
     * Asks the user to confirm canceling the current activity. If no is
     * chosen the user can continue to add or edit the connection. Otherwise
     * the input will be discarded and the activity will be closed.
     */
    private void cancel() {
        if (!connectionValuesChanged) {
            activity.finish();
        } else {
            // Show confirmation dialog to cancel
            new MaterialDialog.Builder(activity)
                    .content(R.string.confirm_discard_connection)
                    .positiveText(getString(R.string.discard))
                    .negativeText(getString(R.string.cancel))
                    .onPositive((dialog, which) -> activity.finish())
                    .onNegative((dialog, which) -> dialog.dismiss())
                    .show();
        }
    }

    @Override
    public void onBackPressed() {
        cancel();
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object o) {
        String value = String.valueOf(o);
        switch (preference.getKey()) {
            case "name":
                if (connection.isNameValid(value)) {
                    connection.setName(value);
                    namePreference.setText(value);
                    namePreference.setSummary(value.isEmpty() ? getString(R.string.pref_name_sum) : value);
                } else {
                    namePreference.setText(connection.getName());
                    SnackbarUtils.sendSnackbarMessage(activity, R.string.pref_name_error_invalid);
                }
                break;
            case "hostname":
                connectionValuesChanged = true;
                if (connection.isIpAddressValid(value)) {
                    connection.setHostname(value);
                    hostnamePreference.setText(value);
                    hostnamePreference.setSummary(value.isEmpty() ? getString(R.string.pref_host_sum) : value);
                } else {
                    hostnamePreference.setText(connection.getHostname());
                    SnackbarUtils.sendSnackbarMessage(activity, R.string.pref_host_error_invalid);
                }
                break;
            case "htsp_port":
                connectionValuesChanged = true;
                try {
                    int port = Integer.parseInt(value);
                    if (connection.isPortValid(port)) {
                        connection.setPort(port);
                        htspPortPreference.setText(String.valueOf(port));
                        htspPortPreference.setSummary(String.valueOf(port));
                    } else {
                        htspPortPreference.setText(String.valueOf(connection.getPort()));
                        SnackbarUtils.sendSnackbarMessage(activity, R.string.pref_port_error_invalid);
                    }
                } catch (NumberFormatException nex) {
                    // NOP
                }
                break;
            case "streaming_port":
                connectionValuesChanged = true;
                try {
                    int port = Integer.parseInt(value);
                    if (connection.isPortValid(port)) {
                        connection.setStreamingPort(port);
                        streamingPortPreference.setText(value);
                        streamingPortPreference.setSummary(getString(R.string.pref_streaming_port_sum, port));
                    } else {
                        streamingPortPreference.setText(String.valueOf(connection.getStreamingPort()));
                        SnackbarUtils.sendSnackbarMessage(activity, R.string.pref_port_error_invalid);
                    }
                } catch (NumberFormatException e) {
                    // NOP
                }
                break;
            case "username":
                connectionValuesChanged = true;
                connection.setUsername(value);
                usernamePreference.setText(value);
                usernamePreference.setSummary(value.isEmpty() ? getString(R.string.pref_user_sum) : value);
                break;
            case "password":
                connectionValuesChanged = true;
                connection.setPassword(value);
                passwordPreference.setText(value);
                passwordPreference.setSummary(value.isEmpty() ? getString(R.string.pref_pass_sum) : getString(R.string.pref_pass_set_sum));
                break;
            case "active_enabled":
                connectionValuesChanged = true;
                // When the connection was set as the new active
                // connection, an initial sync is required
                boolean isActive = Boolean.valueOf(value);
                if (!connection.isActive() && isActive) {
                    connection.setSyncRequired(true);
                    connection.setLastUpdate(0);
                }
                connection.setActive(Boolean.valueOf(value));
                break;
            case "wol_enabled":
                connection.setWolEnabled(Boolean.valueOf(value));
                break;
            case "wol_mac_address":
                if (connection.isWolMacAddressValid(value)) {
                    connection.setWolMacAddress(value);
                    wolMacAddressPreference.setText(value);
                    wolMacAddressPreference.setSummary(value.isEmpty() ? getString(R.string.pref_wol_address_sum) : value);
                } else {
                    wolMacAddressPreference.setText(connection.getWolMacAddress());
                    SnackbarUtils.sendSnackbarMessage(activity, R.string.pref_wol_address_invalid);
                }
                break;
            case "wol_port":
                try {
                    int port = Integer.parseInt(value);
                    if (connection.isPortValid(port)) {
                        connection.setWolPort(port);
                        wolPortPreference.setText(value);
                        wolPortPreference.setSummary(getString(R.string.pref_wol_port_sum, port));
                    } else {
                        wolPortPreference.setText(String.valueOf(connection.getWolPort()));
                        SnackbarUtils.sendSnackbarMessage(activity, R.string.pref_port_error_invalid);
                    }
                } catch (NumberFormatException e) {
                    // NOP
                }
                break;
            case "wol_broadcast_enabled":
                connection.setWolUseBroadcast(Boolean.valueOf(value));
                break;
        }
        return true;
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/settings/SettingsConnectionBaseFragment.kt
package org.tvheadend.tvhclient.ui.features.settings

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.ViewModelProvider
import androidx.preference.EditTextPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.SwitchPreference
import com.afollestad.materialdialogs.MaterialDialog
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.ui.common.interfaces.BackPressedInterface
import org.tvheadend.tvhclient.ui.common.interfaces.ToolbarInterface
import org.tvheadend.tvhclient.util.extensions.sendSnackbarMessage

abstract class SettingsConnectionBaseFragment : PreferenceFragmentCompat(), BackPressedInterface, Preference.OnPreferenceChangeListener {

    lateinit var toolbarInterface: ToolbarInterface
    lateinit var settingsViewModel: SettingsViewModel
    val connectionValidator = ConnectionValidator()

    private lateinit var namePreference: EditTextPreference
    private lateinit var serverUrlPreference: EditTextPreference
    private lateinit var streamingUrlPreference: EditTextPreference
    private lateinit var usernamePreference: EditTextPreference
    private lateinit var passwordPreference: EditTextPreference
    protected lateinit var activeEnabledPreference: SwitchPreference
    private lateinit var wolMacAddressPreference: EditTextPreference
    private lateinit var wolPortPreference: EditTextPreference
    private lateinit var wolEnabledPreference: SwitchPreference
    private lateinit var wolUseBroadcastEnabled: SwitchPreference

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (activity is ToolbarInterface) {
            toolbarInterface = activity as ToolbarInterface
        }

        settingsViewModel = ViewModelProvider(activity as SettingsActivity)[SettingsViewModel::class.java]
        setHasOptionsMenu(true)

        // Get the connectivity preferences for later usage
        namePreference = findPreference("name")!!
        serverUrlPreference = findPreference("server_url")!!
        streamingUrlPreference = findPreference("streaming_url")!!
        usernamePreference = findPreference("username")!!
        passwordPreference = findPreference("password")!!
        activeEnabledPreference = findPreference("active_enabled")!!
        wolEnabledPreference = findPreference("wol_enabled")!!
        wolMacAddressPreference = findPreference("wol_mac_address")!!
        wolPortPreference = findPreference("wol_port")!!
        wolUseBroadcastEnabled = findPreference("wol_broadcast_enabled")!!

        namePreference.onPreferenceChangeListener = this
        serverUrlPreference.onPreferenceChangeListener = this
        streamingUrlPreference.onPreferenceChangeListener = this
        usernamePreference.onPreferenceChangeListener = this
        passwordPreference.onPreferenceChangeListener = this
        activeEnabledPreference.onPreferenceChangeListener = this
        wolEnabledPreference.onPreferenceChangeListener = this
        wolMacAddressPreference.onPreferenceChangeListener = this
        wolPortPreference.onPreferenceChangeListener = this
        wolUseBroadcastEnabled.onPreferenceChangeListener = this
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences_add_connection, rootKey)
    }

    override fun onResume() {
        super.onResume()
        setPreferenceDefaultValues()
    }

    private fun setPreferenceDefaultValues() {
        val name = settingsViewModel.connectionToEdit.name
        namePreference.text = name
        namePreference.summary = if (name.isNullOrEmpty()) getString(R.string.pref_name_sum) else name

        val serverUrl = settingsViewModel.connectionToEdit.serverUrl
        serverUrlPreference.text = serverUrl
        serverUrlPreference.summary = if (serverUrl.isNullOrEmpty()) getString(R.string.pref_server_url_sum) else serverUrl

        val streamingUrl = settingsViewModel.connectionToEdit.streamingUrl
        streamingUrlPreference.text = streamingUrl
        streamingUrlPreference.summary = if (streamingUrl.isNullOrEmpty()) getString(R.string.pref_streaming_url_sum) else streamingUrl

        val username = settingsViewModel.connectionToEdit.username
        usernamePreference.text = username
        usernamePreference.summary = if (username.isNullOrEmpty() || username == "*") getString(R.string.pref_user_sum) else username

        val password = settingsViewModel.connectionToEdit.password
        passwordPreference.text = password
        passwordPreference.summary = if (password.isNullOrEmpty() || password == "*") getString(R.string.pref_pass_sum) else getString(R.string.pref_pass_set_sum)

        activeEnabledPreference.isChecked = settingsViewModel.connectionToEdit.isActive
        wolEnabledPreference.isChecked = settingsViewModel.connectionToEdit.isWolEnabled

        if (!settingsViewModel.connectionToEdit.isWolEnabled) {
            val wolPort = settingsViewModel.connectionToEdit.wolPort
            val wolMacAddress = settingsViewModel.connectionToEdit.wolMacAddress

            wolMacAddressPreference.text = wolMacAddress
            wolMacAddressPreference.summary = if (wolMacAddress.isNullOrEmpty()) getString(R.string.pref_wol_address_sum) else wolMacAddress
            wolPortPreference.text = wolPort.toString()
            wolPortPreference.summary = getString(R.string.pref_wol_port_sum, wolPort)
            wolUseBroadcastEnabled.isChecked = settingsViewModel.connectionToEdit.isWolUseBroadcast
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.save_cancel_options_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_save -> {
                save()
                true
            }
            R.id.menu_cancel -> {
                cancel()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    protected abstract fun save()

    fun getErrorDescription(reason: ValidationFailureReason, urlType: String = ""): String {
        return when (reason) {
            is ValidationFailureReason.NameEmpty -> context?.resources?.getString(R.string.pref_name_error_invalid) ?: ""
            is ValidationFailureReason.NameInvalid -> context?.resources?.getString(R.string.pref_name_error_invalid) ?: ""
            is ValidationFailureReason.UrlEmpty -> "The $urlType url must not be empty"
            is ValidationFailureReason.UrlSchemeMissing -> "The $urlType url must start with http:// or https://"
            is ValidationFailureReason.UrlSchemeWrong -> "The $urlType url must contain http:// or https://"
            is ValidationFailureReason.UrlHostMissing -> "The $urlType url is missing a hostname"
            is ValidationFailureReason.UrlPortMissing -> "The $urlType url is missing a port number"
            is ValidationFailureReason.UrlPortInvalid -> context?.resources?.getString(R.string.pref_port_error_invalid) ?: ""
            is ValidationFailureReason.UrlUnneededCredentials -> "The url must not contain a username or password"
            is ValidationFailureReason.MacAddressEmpty -> context?.resources?.getString(R.string.pref_wol_address_invalid) ?: ""
            is ValidationFailureReason.MacAddressInvalid -> context?.resources?.getString(R.string.pref_wol_address_invalid) ?: ""
        }
    }

    /**
     * Asks the user to confirm canceling the current activity. If no is
     * chosen the user can continue to add or edit the connection. Otherwise
     * the input will be discarded and the activity will be closed.
     */
    private fun cancel() {
        activity?.let { activity ->
            MaterialDialog(activity).show {
                message(R.string.confirm_discard_connection)
                positiveButton(R.string.discard) {
                    if (activity is RemoveFragmentFromBackstackInterface) {
                        activity.removeFragmentFromBackstack()
                    }
                }
                negativeButton(R.string.cancel) {
                    dismiss()
                }
            }
        }
    }

    override fun onBackPressed() {
        cancel()
    }

    override fun onPreferenceChange(preference: Preference, o: Any): Boolean {
        val value = o.toString()
        when (preference.key) {
            "name" -> preferenceNameChanged(value)
            "server_url" -> preferenceUrlChanged(value)
            "streaming_url" -> preferenceStreamingUrlChanged(value)
            "username" -> preferenceUsernameChanged(value)
            "password" -> preferencePasswordChanged(value)
            "active_enabled" -> preferenceEnabledChanged(value)
            "wol_enabled" -> preferenceWolEnabledChanged(value)
            "wol_mac_address" -> preferenceWolMacAddressChanged(value)
            "wol_port" -> preferenceWolPortChanged(value)
            "wol_broadcast_enabled" -> preferenceWolBroadcastChanged(value)
        }
        return true
    }

    private fun preferenceNameChanged(value: String) {
        when (val result = connectionValidator.isConnectionNameValid(value)) {
            is ValidationResult.Success -> {
                settingsViewModel.connectionToEdit.name = value
                namePreference.text = value
                namePreference.summary = value.ifEmpty { getString(R.string.pref_name_sum) }
            }
            is ValidationResult.Failed -> {
                namePreference.text = settingsViewModel.connectionToEdit.name
                context?.sendSnackbarMessage(getErrorDescription(result.reason))
            }
        }
    }

    private fun preferenceUrlChanged(value: String) {
        when (val result = connectionValidator.isConnectionUrlValid(value)) {
            is ValidationResult.Success -> {
                settingsViewModel.connectionToEdit.serverUrl = value
                serverUrlPreference.text = value
                serverUrlPreference.summary = value.ifEmpty { getString(R.string.pref_server_url_sum) }
            }
            is ValidationResult.Failed -> {
                serverUrlPreference.text = settingsViewModel.connectionToEdit.serverUrl
                context?.sendSnackbarMessage(getErrorDescription(result.reason, "connection"))
            }
        }
    }

    private fun preferenceStreamingUrlChanged(value: String) {
        when (val result = connectionValidator.isConnectionUrlValid(value)) {
            is ValidationResult.Success -> {
                settingsViewModel.connectionToEdit.streamingUrl = value
                streamingUrlPreference.text = value
                streamingUrlPreference.summary = value.ifEmpty { getString(R.string.pref_streaming_url_sum) }
            }
            is ValidationResult.Failed -> {
                streamingUrlPreference.text = settingsViewModel.connectionToEdit.streamingUrl
                context?.sendSnackbarMessage(getErrorDescription(result.reason, "playback"))
            }
        }
    }

    private fun preferenceUsernameChanged(value: String) {
        settingsViewModel.connectionToEdit.username = value
        usernamePreference.text = value
        usernamePreference.summary = value.ifEmpty { getString(R.string.pref_user_sum) }
    }

    private fun preferencePasswordChanged(value: String) {
        settingsViewModel.connectionToEdit.password = value
        passwordPreference.text = value
        passwordPreference.summary = if (value.isEmpty()) getString(R.string.pref_pass_sum) else getString(R.string.pref_pass_set_sum)
    }

    private fun preferenceEnabledChanged(value: String) {
        // When the connection was set as the new active
        // connection, an initial sync is required
        val isActive = java.lang.Boolean.valueOf(value)
        if (!settingsViewModel.connectionToEdit.isActive && isActive) {
            settingsViewModel.connectionToEdit.isSyncRequired = true
            settingsViewModel.connectionToEdit.lastUpdate = 0
        }
        settingsViewModel.connectionToEdit.isActive = java.lang.Boolean.valueOf(value)
    }

    private fun preferenceWolEnabledChanged(value: String) {
        settingsViewModel.connectionToEdit.isWolEnabled = java.lang.Boolean.valueOf(value)
    }

    private fun preferenceWolMacAddressChanged(value: String) {
        when (val result = connectionValidator.isConnectionWolMacAddressValid(value)) {
            is ValidationResult.Success -> {
                settingsViewModel.connectionToEdit.wolMacAddress = value
                wolMacAddressPreference.text = value
                wolMacAddressPreference.summary = value.ifEmpty { getString(R.string.pref_wol_address_sum) }
            }
            is ValidationResult.Failed -> {
                wolMacAddressPreference.text = settingsViewModel.connectionToEdit.wolMacAddress
                context?.sendSnackbarMessage(getErrorDescription(result.reason))
            }
        }
    }

    private fun preferenceWolPortChanged(value: String) {
        try {
            val port = Integer.parseInt(value)
            when (val result = connectionValidator.isConnectionWolPortValid(port)) {
                is ValidationResult.Success -> {
                    settingsViewModel.connectionToEdit.wolPort = port
                    wolPortPreference.text = value
                    wolPortPreference.summary = getString(R.string.pref_wol_port_sum, port)
                }
                is ValidationResult.Failed -> {
                    wolPortPreference.text = settingsViewModel.connectionToEdit.wolPort.toString()
                    context?.sendSnackbarMessage(getErrorDescription(result.reason))
                }
            }
        } catch (e: NumberFormatException) {
            // NOP
        }
    }

    private fun preferenceWolBroadcastChanged(value: String) {
        settingsViewModel.connectionToEdit.isWolUseBroadcast = java.lang.Boolean.valueOf(value)
    }
}
